videosink/handler.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2021 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package videosink
+
+import (
+	"bytes"
+	"fmt"
+	"image/jpeg"
+	"log"
+	"mime"
+	"net/http"
+	"net/textproto"
+	"net/url"
+	"sync"
+)
+
+// bufferPool stores reusable []byte instances.
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return []byte(nil)
+	},
+}
+
+type imageConfig struct {
+	format ImageFormat
+}
+
+func (d *Display) configFromQuery(values url.Values) (imageConfig, error) {
+	cfg := imageConfig{
+		format: d.defaultFormat,
+	}
+
+	if value := values.Get("format"); value != "" {
+		if format, err := ImageFormatFromString(value); err != nil {
+			return imageConfig{}, err
+		} else {
+			cfg.format = format
+		}
+	}
+
+	return cfg, nil
+}
+
+type client struct {
+	refresh   chan struct{}
+	terminate chan struct{}
+}
+
+func (d *Display) bufferChangedLocked() {
+	for cfg, buffer := range d.snapshot {
+		//lint:ignore SA6002 buffer is []byte and thus pointer-like
+		bufferPool.Put(buffer)
+		delete(d.snapshot, cfg)
+	}
+
+	for c := range d.clients {
+		select {
+		case c.refresh <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (d *Display) encodeBufferLocked(format ImageFormat) ([]byte, error) {
+	buf := bytes.NewBuffer(bufferPool.Get().([]byte)[:0])
+
+	switch format {
+	case PNG:
+		if err := pngEncoder.get().Encode(buf, d.buffer); err != nil {
+			return nil, err
+		}
+
+	case JPEG:
+		if err := jpeg.Encode(buf, d.buffer, &jpegOptions); err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, fmt.Errorf("unhandled image format %s", format)
+	}
+
+	return buf.Bytes(), nil
+}
+
+func (d *Display) grabSnapshot(cfg imageConfig) ([]byte, error) {
+	var err error
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	encoded, ok := d.snapshot[cfg]
+	if !ok {
+		encoded, err = d.encodeBufferLocked(cfg.format)
+		if err != nil {
+			return nil, err
```

Comment:
The only case where it could fail if if there's an encoder configuration error or the display type is corrupted. I think you can just return nil here.

Reply:
I think even a panic here would be fine and still not hide the problem. `grabSnapshot` is called from the HTTP request handler for which Go's HTTP server handles panics. What do you think?

Reply:
Generally not fond of panic but I think it's a fine choice here.

Reply:
Done.